support/refc/_datatypes.h

```diff
@@ -50,15 +45,47 @@ typedef struct {
   // followed by type-specific payload.
 } Value;
 
-typedef struct {
-  Value_header header;
-  uint8_t ui8;
-} Value_Bits8;
+/* expcted at least 4bytes for Value_header alignment. */
+#define idris2_vp_is_unboxed(p) ((uintptr_t)(p)&3)
 
-typedef struct {
-  Value_header header;
-  uint16_t ui16;
-} Value_Bits16;
+#define idris2_vp_int_shift                                                    \
+  ((sizeof(uintptr_t) >= 8 && sizeof(Value *) >= 8) ? 32 : 16)
+
+#define idris2_vp_to_Bits64(p) (((Value_Bits64 *)(p))->ui64)
+
+#if !defined(UINTPTR_WIDTH)
+#define idris2_vp_to_Bits32(p)                                                 \
+  ((idris2_vp_int_shift == 16)                                                 \
+       ? (((Value_Bits32 *)(p))->ui32)                                         \
+       : ((uint32_t)((uintptr_t)(p) >> idris2_vp_int_shift)))
+#define idris2_vp_to_Int32(p)                                                  \
+  ((idris2_vp_int_shift == 16)                                                 \
+       ? (((Value_Int32 *)(p))->i32)                                           \
+       : ((int32_t)((uintptr_t)(p) >> idris2_vp_int_shift)))
+
+#elif UINTPTR_WIDTH >= 64
+#define idris2_vp_to_Bits32(p)                                                 \
+  ((uint32_t)((uintptr_t)(p) >> idris2_vp_int_shift))
+#define idris2_vp_to_Int32(p): ((int32_t)((uintptr_t)(p) >> idris2_vp_int_shift))
+
+#elif UINTPTR_WIDTH >= 32
+#define idris2_vp_to_Bits32(p) (((Value_Bits32 *)(p))->ui32)
+#define idris2_vp_to_Int32(p) (((Value_Int32 *)(p))->i32)
+
+#else
+#error "unsupported uintptr_t width"
+#endif
+
+#define idris2_vp_to_Bits16(p)                                                 \
+  ((uint16_t)((uintptr_t)(p) >> idris2_vp_int_shift))
+#define idris2_vp_to_Bits8(p) ((uint8_t)((uintptr_t)(p) >> idris2_vp_int_shift))
+#define idris2_vp_to_Int64(p) (((Value_Int64 *)(p))->i64)
```

Comment:
Could you explain why you are still boxing 64-bit values even in the case when `UINTPTR_WIDTH >= 64`?

Reply:
Since bit 0 of the pointer is used as a flag, the maximum number of bits that can be unboxed is 63. (My code also took away bit 1, so 62 bits is the limit.) And since we are not thinking about 128-bit CPUs, int64 will not be able to unbox.

Reply:
Ok. And could you comment why is this flag needed?

Reply:
This flag is used to identify whether a pointer is actually a Value* or an integer.

RefC does not have complete static tracking of type information, so types are identified at runtime using Value_Header's tag field. However, Int that are pretending to be pointers cannot have that tag, so use that flag to identify them first.
Of course, this flag is not used if it is clear that Value* is actually an Int. But places like newReference/removeReference require this flag.

Reply:
In fact. this unboxing is a classic idea, but It seems to be difficult to understand, so I added a comment to the source.

Reply:
> But places like newReference/removeReference require this flag

Is it possible in principle to not to call to `removeReference` for unboxed values when still having the typing information?

Reply:
Partially, yes. I guess.
If AVar had type information, it might reduce the number of calls to removeReference, and integers shouldn't even have to pretend to be pointers. Maybe until we bind it somewhere or use believe_me.

In the case of closures and constructors, static information is not very useful for values already bound to them, and eventually we have to rely on dynamic information. However, we can put the information in the header part of the closure instead of the pointer bit 0.